src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -945,6 +960,58 @@ internal bool CalculateUseSiteDiagnostic(ref DiagnosticInfo result)
             return false;
         }
 
+#nullable enable
+        protected static UnmanagedCallersOnlyAttributeData DecodeUnmanagedCallersOnlyAttributeData(CSharpAttributeData attribute, Location? location, DiagnosticBag? diagnostics)
+        {
+            Debug.Assert((location == null) == (diagnostics == null));
+            ImmutableHashSet<INamedTypeSymbolInternal>? callingConventionTypes = null;
+            bool isValid = true;
+            if (!attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    if (callingConventionTypes != null)
+                    {
+                        isValid = false;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        var builder = PooledHashSet<INamedTypeSymbolInternal>.GetInstance();
```

Comment:
>var builder [](start = 24, length = 11)

Should the builder be outside the loop?

If not, consider skipping the builder if there are only zero or one types:
```
callingConventionTypes = ImmutableHasSet<INamedTypeSymbol>.Empty;
if (...)
{
    isValid = false;
}
else
{
    callingConventionTypes = callingConventionTypes.Add(callConvType);
}
```

Reply:
I considered it, but the only thing that would affect would be error scenarios and I felt this way was cleaner.

---
In reply to: [482599236](https://github.com/dotnet/roslyn/pull/47311#discussion_r482599236) [](ancestors = 482599236)